File: component_demo1/src/main/java/com/example/component_demo1/http/WeatherApi.java
package com.example.component_demo1.http;


import java.io.IOException;
import java.util.HashMap;

import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;
import org.json.JSONException;
import retrofit2.Call;
import retrofit2.http.GET;
import retrofit2.http.QueryMap;

/**
 * author: banketree
 * created on: 2019/6/12
 * description:
 * 天气预报
 */
public class WeatherApi extends ApiService {
    private final String key = "4e34d358d9aa062b2c46afd627084f85";
    private final String BASE_URL = "https://restapi.amap.com";

    @Override
    protected String getUrl() {
        return BASE_URL;
    }

    @Override
    protected Class<? extends Object> getService() {
        return WeatherServiceApi.class;
    }

    //添加头信息()
    @Override
    protected Response okhttpInterceptor(Interceptor.Chain chain) throws IOException {
        Request.Builder builder = chain.request()
                .newBuilder()
                .removeHeader("Content-Type")
                .addHeader("Content-Type", "application/x-www-form-urlencoded;charset=UTF-8")
                .addHeader("charset", "UTF-8"); //登录前 必传

        Request request = builder.build();
        return chain.proceed(request);
    }

    //通过ip 获取城市编码
    private void getCityCodeByIp(HttpCallback callback) {
        HashMap<String, String> hashMap = new HashMap<>();
        hashMap.put("output", "json");
        hashMap.put("key", key);
        doNet("getCityCodeByIp", hashMap, callback);
    }

    //获取天气预报
    //步骤1：通过ip 获取城市信息
    //步骤2：通过城市编码 获取天气信息
    public void getWeather(final HttpCallback callback) {
        getCityCodeByIp(new HttpCallback<String>() {
            @Override
            public void onSucess(String respBase) {
                super.onSucess(respBase);
            }

            @Override
            public void onFaile(Exception ex) {
                super.onFaile(ex);
                if (callback != null) callback.onFaile(ex);
            }

            @Override
            public void onCancel() {
                super.onCancel();
                if (callback != null) callback.onCancel();
            }

            @Override
            protected String getBean(retrofit2.Response response) throws JSONException {
                return super.getBean(response);
            }
        });
    }

    //    设置专属
    public interface WeatherServiceApi {
        /**
         * 通过ip 查城市信息
         */
        @GET("/v3/ip")
        Call<String> getCityCodeByIp(@QueryMap HashMap<String, String> map);
    }
}

File: component_demo1/src/main/java/com/example/component_demo1/http/WeatherApi.kt
package com.example.component_demo1.http

import com.example.base_fun.http.RetrofitFactory
import com.example.component_demo1.ui.mvvm.RespWeather
import me.jessyan.retrofiturlmanager.RetrofitUrlManager
import okhttp3.OkHttpClient
import retrofit2.Response
import java.util.HashMap

import retrofit2.http.GET
import retrofit2.http.Headers
import retrofit2.http.QueryMap
import java.util.concurrent.TimeUnit

/**
 * author: banketree
 * created on: 2019/6/12
 * description:
 * 天气预报
 */
class WeatherApi : RetrofitFactory() {
    companion object {
        private const val key = "4e34d358d9aa062b2c46afd627084f85"
        private const val BASE_URL = "https://restapi.amap.com"
    }

    override fun getUrl(): String = BASE_URL

    override fun handleOkHttpClient(builder: OkHttpClient.Builder) {
        super.handleOkHttpClient(builder) //有默认参数

        builder.apply {
            connectTimeout(50, TimeUnit.SECONDS)
            readTimeout(50, TimeUnit.SECONDS)
            writeTimeout(50, TimeUnit.SECONDS)
        }
    }

    //通过ip 获取城市编码
    suspend fun getCityCodeByIpByString(): Response<String> {
        val hashMap = HashMap<String, String>()
        hashMap["output"] = "json"
        hashMap["key"] = key
        return stringService(WeatherServiceApi::class.java).getCityCodeByIp(hashMap)
    }

    suspend fun getCityCodeByIpByGson(): Response<RespWeather> {
        val hashMap = HashMap<String, String>()
        hashMap["output"] = "json"
        hashMap["key"] = key
        return gsonService(WeatherServiceApi::class.java).getCityCodeByIp2(hashMap)
    }


    //    设置专属
    interface WeatherServiceApi {
        /**
         * 通过ip 查城市信息
         */
        @Headers("domain-test3:$BASE_URL")  // 加上 Domain-Name header
        @GET("/v3/ip")
        suspend fun getCityCodeByIp(@QueryMap map: HashMap<String, String>): Response<String>

        @GET("/v3/ip")
        suspend fun getCityCodeByIp2(@QueryMap map: HashMap<String, String>): Response<RespWeather>
    }
}
